controllers/pipelinegates/pipelinegate_controller.go

```diff
@@ -0,0 +1,376 @@
+			return ctrl.Result{}, err
+		}
+		log.V(1).Info("Updated state of CR on first reconcile.", "Namespace", gate.Namespace, "Name", gate.Name, "ID", gate.Spec.ID, "SyncedState", gate.Spec.SyncedState, "LastSyncedAt", gate.Spec.LastSyncedAt, "State", gate.Status.State)
+		return ctrl.Result{}, nil
+	}
+
+	// PENDING
+	if gate.Status.State != nil && *gate.Status.State == pipelinesv1alpha1.GateState(console.GateStatePending) {
+		log.V(2).Info("Reconciling PENDING gate.", "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State)
+
+		if (gate.Status.JobRef == nil || *gate.Status.JobRef == console.NamespacedName{}) {
+			log.V(2).Info("Gate doesn't have a JobRef, this is a new gate or a re-run.", "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State)
+
+			jobName := fmt.Sprintf("%s-%s", gate.Name, uuid.New().String())
+			jobRef := console.NamespacedName{Name: jobName, Namespace: gate.Namespace}
+			job := r.generateJob(ctx, log, *gate.Spec.GateSpec.JobSpec, jobRef)
+			if err := ctrl.SetControllerReference(gate, job, r.Scheme); err != nil {
+				log.Error(err, "Error setting ControllerReference for Job.")
+				return ctrl.Result{}, err
+			}
+
+			// reconcile job
+
+			log.V(2).Info("Creating new job for gate.", "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "jobRef", jobRef)
+			_, err := Job(ctx, r.Client, job, log)
+			if err != nil {
+				log.Error(err, "Error reconciling Job.", "JobName", job.Name, "Namespace", job.Namespace)
+				return ctrl.Result{}, err
+			}
+
+			gateState := pipelinesv1alpha1.GateState(console.GateStatePending)
+			gate.Status.State = &gateState
+			gate.Status.JobRef = &jobRef
+			if err := r.Status().Update(ctx, gate); err != nil {
+				log.Error(err, "Failed to update PipelineGate status at CR.")
+				return ctrl.Result{}, err
+			}
+
+			// try to update gate at console
+			attributeState := console.GateStatePending
+			updateAttributes := console.GateUpdateAttributes{State: &attributeState, Status: &console.GateStatusAttributes{JobRef: gate.Status.JobRef}}
+			if err := r.ConsoleClient.UpdateGate(gate.Spec.ID, updateAttributes); err != nil {
+				log.Error(err, "Failed to update PipelineGate status to console")
+			}
+
+			log.V(1).Info("Created new job for gate and updated status.", "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "jobRef", *gate.Status.JobRef)
+			return ctrl.Result{}, nil
+		} else {
+			log.V(2).Info("Gate has a JobRef, checking Job status.", "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "jobRef", *gate.Status.JobRef)
+			job := r.generateJob(ctx, log, *gate.Spec.GateSpec.JobSpec, *gate.Status.JobRef)
+			if err := ctrl.SetControllerReference(gate, job, r.Scheme); err != nil {
+				log.Error(err, "Error setting ControllerReference for Job.")
+				return ctrl.Result{}, err
+			}
+			// reconcile job, creates a new one or gets the old one
+			reconciledJob, err := Job(ctx, r.Client, job, log)
```

Comment:
We should create reconciler for jobs and change the gate status from it once job is finished. We could watch only for jobs that are connected with gates. Then this reconciler would be taking care only of gates (and creating job).

Reply:
that sounds like an antipattern, tbh, we're already modifying the state from two places